### PR TITLE
core: Allow graph regions in printer

### PR DESCRIPTION
**Pull request by math-fehr**
This PR allows graph regions to be printed by the `Printer`.
Since there is no way of knowing if the user is currently printing operations in a graph region, this means that we cannot print operands with a `%<UNKNOWN>` when they haven't been printed yet.
Thus, printing an operation that has free operands will not fail anymore.

Note that this means that some dominance bugs are not catched anymore. These should be handled with a proper domination check in the verifier later on.

### CODE REVIEW

**Review by tobiasgrosser — APPROVED** [review]
Very cool!

File: tests/test_ir.py [hunk 1]
```diff
@@ -415,7 +415,7 @@ def test_op_custom_verify_is_done_last():
     assert e.value.args[0] != "Custom Verification Check"
     assert (
         e.value.args[0]
-        == 'test.custom_verify_op operation does not verify\n\n"test.custom_verify_op"(%<UNKNOWN>) : (i32) -> ()\n\n'
+        == 'test.custom_verify_op operation does not verify\n\n"test.custom_verify_op"(%0) : (i32) -> ()\n\n'
```
Comment (webmiche) [inline tests/test_ir.py:418]: Is their something like `{{.*}}` for this case? Maybe just check prefix and postfix or something?

**Review by webmiche — APPROVED** [review]
Nice, I think that's actually the better design anyhow. This will remove a lot of annoyance for people trying to print their ops :+1: 

Could you open an issue about dominance? Just to keep record of things that are still open. I think this could actually be a cool coding challenge for compiler beginners :thinking:

File: tests/test_ir.py (hunk 1 above)
Comment (math-fehr) [reply tests/test_ir.py]: Is this better? We don't actually need to check for the actual message, just that the IRDL check got hit

**Review by superlopuh — APPROVED** [review]
LGTM! Surprised that graph regions don't have some annotation on them to decide which behaviour we want

**Comment by math-fehr** [comment]
> LGTM! Surprised that graph regions don't have some annotation on them to decide which behaviour we want

This is going to be added after in IRDL, but graph regions do not have annotations at the IR level.
It is a trait that operations carry, and it only influence the domination check in the verifier, but the parsing allows everything roughly.

**Review by georgebisbas — APPROVED** [review]

**Pull request merged**